### ISSUE TITLE
tls: handshake failure

### ISSUE THREAD

**Issue by adrianosantos** [sijms/go-ora#219]
I'm trying to connect to Oracle DB with SSL using wallet and I'm getting the pls handshake failure error.

func ConnectDB(dbParams map[string]string) {
	connectionString := "oracle://" + dbParams["username"] + ":" + dbParams["password"] + "@" + dbParams["server"] + ":" + dbParams["port"] + "/" + dbParams["service"]
	if val, ok := dbParams["walletLocation"]; ok && val != "" {
		connectionString += "?TRACE FILE=trace.log&SSL=enable&Verify SSL=false&wallet=" + url.QueryEscape(dbParams["walletLocation"])
	}
	 
	db, err := sql.Open("oracle", connectionString)
	if err != nil {
		panic(fmt.Errorf("error in sql.Open: %w", err))
	}
	

	defer func() {
		err = db.Close()
		if err != nil {
			fmt.Println("Can't close connection: ", err)
		}
	}()
    
	err = db.Ping()
	if err != nil {
		panic(fmt.Errorf("error pinging db: %w", err))
	}
}

When I execute the code the db ping fails.

panic: error pinging db: remote error: tls: handshake failure

**Comment by sijms** [comment]
please confirm that your certificate in wallet file
also the option you use is incorrect: ssl=enable&ssl verify=false
after that if still have problem
can I know your server version and go-ora version

**Comment by adrianosantos** [comment]
Oracle Version is 19c
go-ora version 2
I made the suggested change and it is still not working. When I try to connect I get the following error:

panic: can't open connection: remote error: tls: handshake failure

**Comment by adrianosantos** [comment]
urlOptions := map[string]string{
		"trace file":        "./logs/trace.log",
		"ssl verify":        "false",
		"wallet":            dbParams["walletLocation"],
		"ssl":               "true",
	}
	databaseURL := go_ora.BuildUrl(dbParams["server"], 1525, dbParams["service"], dbParams["username"], dbParams["password"], urlOptions)
	fmt.Printf("databaseURL: %v\n", databaseURL)
	conn, err := go_ora.NewConnection(databaseURL)
	if err != nil {
		panic(fmt.Errorf("can't create connection:  %w", err))
	}
	err = conn.Open()
	if err != nil {
		panic(fmt.Errorf("can't open connection: %w", err))

	}

**Comment by adrianosantos** [comment]
@sijms Could you please reopen this issue?

**Comment by sijms** [comment]
oracle server: 19.3c
connection using TCPS
this is trace.log of the connection
```
2023-02-03T14:00:18.7998: Connect
2023-02-03T14:00:18.8002: using: localhost:2484 ..... [SUCCEED]
2023-02-03T14:00:18.8002: Using SSL/TLS
2023-02-03T14:00:18.8002: Open :(DESCRIPTION=(ADDRESS=(PROTOCOL=tcps)(HOST=localhost)(PORT=2484))(CONNECT_DATA=(SERVICE_NAME=lab)(CID=(PROGRAM=/home/samy/Projects/Go/TestOracle2/bin/crud_example)(HOST=manager-pc)(USER=samy))))
2023-02-03T14:00:18.8002: 
Write packet:
00000000  01 08 00 00 01 00 00 00  01 3d 01 2c 0c 01 ff ff  |.........=.,....|
00000010  ff ff 4f 98 00 00 00 01  00 c2 00 46 00 00 00 00  |..O........F....|
00000020  01 01 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000030  00 00 00 00 00 00 00 00  00 00 00 20 00 00 00 20  |........... ... |
00000040  00 00 00 00 00 00 28 44  45 53 43 52 49 50 54 49  |......(DESCRIPTI|
00000050  4f 4e 3d 28 41 44 44 52  45 53 53 3d 28 50 52 4f  |ON=(ADDRESS=(PRO|
00000060  54 4f 43 4f 4c 3d 74 63  70 73 29 28 48 4f 53 54  |TOCOL=tcps)(HOST|
00000070  3d 6c 6f 63 61 6c 68 6f  73 74 29 28 50 4f 52 54  |=localhost)(PORT|
00000080  3d 32 34 38 34 29 29 28  43 4f 4e 4e 45 43 54 5f  |=2484))(CONNECT_|
00000090  44 41 54 41 3d 28 53 45  52 56 49 43 45 5f 4e 41  |DATA=(SERVICE_NA|
000000a0  4d 45 3d 6c 61 62 29 28  43 49 44 3d 28 50 52 4f  |ME=lab)(CID=(PRO|
000000b0  47 52 41 4d 3d 2f 68 6f  6d 65 2f 73 61 6d 79 2f  |GRAM=/home/samy/|
000000c0  50 72 6f 6a 65 63 74 73  2f 47 6f 2f 54 65 73 74  |Projects/Go/Test|
000000d0  4f 72 61 63 6c 65 32 2f  62 69 6e 2f 63 72 75 64  |Oracle2/bin/crud|
000000e0  5f 65 78 61 6d 70 6c 65  29 28 48 4f 53 54 3d 6d  |_example)(HOST=m|
000000f0  61 6e 61 67 65 72 2d 70  63 29 28 55 53 45 52 3d  |anager-pc)(USER=|
00000100  73 61 6d 79 29 29 29 29                           |samy))))|
2023-02-03T14:00:19.5180: 
Read packet:
00000000  00 2d 00 00 02 00 00 00  01 3d 0c 01 00 00 00 00  |.-.......=......|
00000010  01 00 00 00 00 2d 41 01  00 00 00 00 00 00 00 00  |.....-A.........|
00000020  00 00 20 00 00 20 00 00  00 00 00 00 00           |.. .. .......|
2023-02-03T14:00:19.5181: Handshake Complete
2023-02-03T14:00:19.5181: Advance Negotiation
2023-02-03T14:00:19.5181: 
Write packet:
00000000  00 00 00 9d 06 00 00 00  00 00 de ad be ef 00 93  |................|
00000010  0b 20 02 00 00 04 00 00  04 00 03 00 00 00 00 00  |. ..............|
00000020  04 00 05 0b 20 02 00 00  08 00 01 00 00 10 1c 66  |.... ..........f|
00000030  ec 28 ea 00 12 00 01 de  ad be ef 00 03 00 00 00  |.(..............|
00000040  04 00 04 00 01 00 02 00  03 00 01 00 03 00 00 00  |................|
00000050  00 00 04 00 05 0b 20 02  00 00 02 00 03 e0 e1 00  |...... .........|
00000060  02 00 06 fc ff 00 02 00  03 00 00 00 00 00 04 00  |................|
00000070  05 0b 20 02 00 00 05 00  01 00 02 0f 10 11 00 01  |.. .............|
00000080  00 02 01 00 03 00 02 00  00 00 00 00 04 00 05 0b  |................|
00000090  20 02 00 00 06 00 01 00  01 03 04 05 06           | ............|
2023-02-03T14:00:19.5183: 
Read packet:
00000000  00 00 00 7f 06 20 00 00  00 00 de ad be ef 00 75  |..... .........u|
00000010  00 00 00 00 00 04 00 00  04 00 03 00 00 00 00 00  |................|
00000020  04 00 05 13 00 00 00 00  02 00 06 00 1f 00 0e 00  |................|
00000030  01 de ad be ef 00 03 00  00 00 02 00 04 00 01 00  |................|
00000040  01 00 02 00 00 00 00 00  04 00 05 13 00 00 00 00  |................|
00000050  02 00 06 fb ff 00 02 00  02 00 00 00 00 00 04 00  |................|
00000060  05 13 00 00 00 00 01 00  02 00 00 03 00 02 00 00  |................|
00000070  00 00 00 04 00 05 13 00  00 00 00 01 00 02 00     |...............|
2023-02-03T14:00:19.5184: TCP Negotiation
2023-02-03T14:00:19.5184: 
Write packet:
00000000  00 00 00 1c 06 00 00 00  00 00 01 06 00 4f 72 61  |.............Ora|
00000010  63 6c 65 43 6c 69 65 6e  74 47 6f 00              |cleClientGo.|
2023-02-03T14:00:19.5185: 
Read packet:
00000000  00 00 00 bf 06 00 00 00  00 00 01 06 00 78 38 36  |.............x86|
00000010  5f 36 34 2f 4c 69 6e 75  78 20 32 2e 34 2e 78 78  |_64/Linux 2.4.xx|
00000020  00 30 02 01 00 00 00 64  00 00 00 60 01 24 0f 05  |.0.....d...`.$..|
00000030  0b 0c 03 0c 0c 05 04 05  0d 06 09 07 08 05 05 05  |................|
00000040  05 05 0f 05 05 05 05 05  0a 05 05 05 05 05 04 05  |................|
00000050  06 07 08 08 23 47 23 47  08 11 23 08 11 41 b0 47  |....#G#G..#..A.G|
00000060  00 83 02 30 07 d0 03 00  00 00 00 00 00 00 00 00  |...0............|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000080  00 00 00 00 00 00 00 00  00 00 00 00 2a 06 01 01  |............*...|
00000090  01 6f 01 01 0c 01 01 01  01 01 01 01 7f ff 03 0e  |.o..............|
000000a0  03 03 01 01 ff 01 ff ff  01 0b 01 01 ff 01 06 0a  |................|
000000b0  e2 01 7f 05 00 0f 0d 07  02 01 00 01 18 00 7f     |...............|
2023-02-03T14:00:19.5186: Server Charset: 560
2023-02-03T14:00:19.5186: Server National Charset: 2000
```

**Comment by sijms** [comment]
when you create oracle wallet you need to pass host name like this
```shell
# add server certificate
orapki wallet add -wallet "/opt/oracle/oradata/server2" -pwd @Password -dn "CN=localhost" -keysize 1024 -self_signed -validity 3650

# add client certificate
orapki wallet add -wallet "/opt/oracle/oradata/client2" -pwd @Password -dn "CN=user-pc" -keysize 1024 -self_signed -validity 3650
```

I face a connection problem when I change my pc host name and connection with TCPS stop working until I recreate wallets with new host name

**Comment by sijms** [comment]
the discussion about SSL connectivity is now in issue #299

**Comment by sijms** [comment]
I fix the issue in v2.5.27
